Carlos Logon component

Cema2019 opened the pull request:
## Hello team,

I've completed my Logon component to connect the app to the Express.js server and Atlas's DB.

I had first tested to fetch the appointment's JSON upon successful sign in as show in this image.

![image](https://github.com/user-attachments/assets/ca310075-e224-4680-9eb1-486991825cb7)

Since Appointment component is not available yet, it will go to 404 page.

![image](https://github.com/user-attachments/assets/d276f345-ab0b-4a41-bcff-f3720ddd7b57)

It's all been written in the FormsWrapper.tsx file which is the one managing the state for the sign-in and sign-up functions. 

Please review and provide feedback as needed.

Cema2019 committed:
Create sign up button, import its material icon and center horizontally

Cema2019 committed:
Convert into a client component

Cema2019 committed:
Convert into a client component

Cema2019 committed:
Finish importing required components

Cema2019 committed:
Create states for login component and login form

Cema2019 committed:
Create states for login component and login form

Cema2019 committed:
Add manual validation before testing with Yup and Formik

Cema2019 committed:
Add manual validation before testing with Yup and Formik

Cema2019 committed:
Form submission event

Cema2019 committed:
Form submission event

Cema2019 committed:
Refactor code to toggle Welcome message with the Login component

Cema2019 committed:
Install axios and update backend port for api conection

Cema2019 committed:
Install axios and update backend port for api conection

Cema2019 committed:
Test connection to database and log appointments to developer console after successful signin

Cema2019 committed:
Test connection to database and log appointments to developer console after successful signin

Cema2019 committed:
Replace changes suggested from develop branch

Cema2019 committed:
Merge branch 'develop' into carlos-signin-component

Cema2019 committed:
Update package-lock.json after merge and npm install

Cema2019 committed:
Update package-lock.json after merge and npm install

Cema2019 committed:
Configure connection to backend and db. Import AxioError library to catch, manage errors to display to console and UI

Cema2019 committed:
Configure connection to backend and db. Import AxioError library to catch, manage errors to display to console and UI

Cema2019 committed:
Remove appointment json test and configured router.push to /appointment which is yet to be created

Cema2019 committed:
Remove appointment json test and configured router.push to /appointment which is yet to be created

Cema2019 committed:
Remove axios library and its dependencies

Cema2019 committed:
Remove axios library and its dependencies

Cema2019 committed:
Refactor code to implement fetch library instead of axios as intructed by team

Cema2019 committed:
Refactor code to implement fetch library instead of axios as intructed by team

Cema2019 committed:
Uninstall axios library from project

Ralexs0096 commented on src/app/components/Forms/FormsWrapper.tsx:
```diff
@@ -6,12 +6,49 @@ import SignInForm from './SignInForm';
-    console.log('Sign In:', values);
+  const handleSignIn = async (values: SignInFormValues) => {
+    try {
+      const response = await fetch('http://localhost:4000/auth/signin', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({
+          email: values.email,
+          password: values.password,
+        }),
+        cache: 'no-store', 
+      });
+
+      if (!response.ok) {
+        const errorData = await response.json();
+        throw new Error(errorData.message || 'Sign-in failed');
+      }
```
it would be good to create a fetchAPI method or something similar, that already has internally the baseURL, which we can take from an environment variable, and that receives only the url, the method and the body if it is required. this will allow to reuse the method between all the application and make it more readable.

Cema2019 replied:
Thanks for the feedback. 
I'm currently working on implementing the suggested changes.

Cema2019 committed:
Create utils folder and fetchAPI function to reuse in the components

Cema2019 committed:
Create utils folder and fetchAPI function to reuse in the components

Cema2019 committed:
Refactor fetchAPI function and relocate the utils folder inside src

Cema2019 committed:
Refactor fetchAPI function and relocate the utils folder inside src

Cema2019 committed:
Update .gitignore to allow commits in .env.example file

Cema2019 committed:
Update .gitignore to allow commits in .env.example file

Cema2019 committed:
Create env example file for API calls

Cema2019 committed:
Create env example file for API calls

Cema2019 commented:
Hello team.

Updates requested are already made and pushed to branch.
The utils folder was created as permitted to host the fetchAPI function which will be reusable.
Sign in attempts were successful in redirecting to the Appointment page.
I also updated .gitgnore as instructed list to allow .env.example file for the repo.

Please review PR.

Ralexs0096 commented on src/app/components/Forms/FormsWrapper.tsx:
```diff
@@ -6,12 +6,36 @@ import SignInForm from './SignInForm';
 import SignUpForm from './SignUpForm';
 import { SignInFormValues, SignUpFormValues } from './Forms.types';
```
non-blocking:
```suggestion
import type { SignInFormValues, SignUpFormValues } from './Forms.types';
```

Ralexs0096 replied:
this change helps the TS compiler to run faster.

Ralexs0096 commented on src/utils/api.ts:
```diff
@@ -0,0 +1,20 @@
+export const fetchAPI = async (endpoint: string, method: string, body?: object) => {
+    const baseURL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:4000';
```
Using `string` allows the function consumer to pass any value of type string. We can force the developer to pass only valid options using this union type. 
```suggestion

type HTTPMethod = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'OPTIONS' | 'HEAD';

export const fetchAPI = async (
  endpoint: string,
  method: HTTPMethod,
  body?: object
) => {
```

Cema2019 committed:
Update src/app/components/Forms/FormsWrapper.tsx

Co-authored-by: Alex Ruiz <49172583+Ralexs0096@users.noreply.github.com>

Cema2019 committed:
Update src/app/components/Forms/FormsWrapper.tsx

Co-authored-by: Alex Ruiz <49172583+Ralexs0096@users.noreply.github.com>

Cema2019 committed:
Update src/utils/api.ts

Add HTTPMethod union type to fetchAPI for improved type safety per review feedback.

Co-authored-by: Alex Ruiz <49172583+Ralexs0096@users.noreply.github.com>

Cema2019 committed:
Update src/utils/api.ts

Add HTTPMethod union type to fetchAPI for improved type safety per review feedback.

Co-authored-by: Alex Ruiz <49172583+Ralexs0096@users.noreply.github.com>

Cema2019 commented:
Team, 

Suggestions have been implemented to the branch. 
Please review further.

Cema2019 committed:
Add mistakenly deleted variable

Cema2019 committed:
Add mistakenly deleted variable

Cema2019 committed:
Rename component for redirection

Cema2019 committed:
Rename component for redirection

Cema2019 committed:
Create appointments folder for the new route

Cema2019 committed:
Create appointments folder for the new route

Cema2019 committed:
Add picture for dummy appointments component

Cema2019 committed:
Add picture for dummy appointments component

Cema2019 committed:
Create component's coding

Cema2019 committed:
Create component's coding

Cema2019 committed:
Added headers parameter and credentials: 'include'

Cema2019 committed:
Added headers parameter and credentials: 'include'

Cema2019 committed:
Add server-actions file to manage sign in and set HTTP-only cookie for security. Logout deletes cookie

Cema2019 committed:
Add server-actions file to manage sign in and set HTTP-only cookie for security. Logout deletes cookie

Cema2019 committed:
Create AuthContext to manage global authentication state

Cema2019 committed:
Create AuthContext to manage global authentication state

Cema2019 committed:
Create route to endpoint to retrieve the HTTP-only token cookie set by signInAction

Cema2019 committed:
Create route to endpoint to retrieve the HTTP-only token cookie set by signInAction

Cema2019 committed:
Create file to enable client-side components to make authenticated API calls by fetching the token from /api/auth/token

Cema2019 committed:
Create file to enable client-side components to make authenticated API calls by fetching the token from /api/auth/token

Cema2019 committed:
Export HTTPMethod to allow import on utils/client-api.ts

Cema2019 committed:
Export HTTPMethod to allow import on utils/client-api.ts

Cema2019 committed:
Add component to check if the user is authenticated before rendering its children

Cema2019 committed:
Add component to check if the user is authenticated before rendering its children

Cema2019 committed:
Refactor component to work with server actions

Cema2019 committed:
Refactor component to work with server actions

Cema2019 committed:
Remove unnecessary imports

Cema2019 committed:
Remove unnecessary imports

Cema2019 committed:
Add type for isSubmitting props

Cema2019 committed:
Add type for isSubmitting props

Cema2019 committed:
Add isSubmitting props

Cema2019 committed:
Add isSubmitting props

Cema2019 committed:
Wrap layout component with AuthProvider

Cema2019 committed:
Wrap layout component with AuthProvider

Cema2019 committed:
Update server-actions

Cema2019 committed:
Update server-actions

Cema2019 committed:
Make updates on useActionState hook

Cema2019 committed:
Make updates on useActionState hook

Cema2019 committed:
Move auth/token/route file to app/api for route convention

Cema2019 committed:
Move auth/token/route file to app/api for route convention

Cema2019 committed:
Wrap component inside ProtectedRoute

Cema2019 committed:
Wrap component inside ProtectedRoute

Cema2019 committed:
Enhance AuthProvider to expose isLoading and handle token fetch errors

Cema2019 committed:
Enhance AuthProvider to expose isLoading and handle token fetch errors

Cema2019 committed:
Update ProtectedRoute to use isLoading and isChecked for reliable auth checks.

Cema2019 committed:
Update ProtectedRoute to use isLoading and isChecked for reliable auth checks.

Cema2019 committed:
Work on useActionState error in FormsWrapper for sign-in redirect

Cema2019 committed:
Work on useActionState error in FormsWrapper for sign-in redirect

Cema2019 committed:
Add logout button

Cema2019 committed:
Add logout button

Cema2019 committed:
Remove fetchAPI('/auth/logout', 'POST') from logout and refactor to router.replace()

Cema2019 committed:
Remove fetchAPI('/auth/logout', 'POST') from logout and refactor to router.replace()

Cema2019 committed:
Remove unnecessary import

Cema2019 committed:
Remove unnecessary import

Cema2019 committed:
Add useRouter and useAuth’s isAuthenticated, isLoading

Cema2019 committed:
Add useRouter and useAuth’s isAuthenticated, isLoading

Ralexs0096 commented on src/utils/client-api.ts:
```diff
@@ -0,0 +1,18 @@
+  if (!response.ok) throw new Error('No token available');
+  const { token } = await response.json();
+  const headers = {
+    ...(token && { 'x-access-token': token }),
```
👍🏻

Ralexs0096 commented on src/app/components/ProtectedRoute.tsx:
```diff
@@ -0,0 +1,29 @@
+export function ProtectedRoute({ children }: { children: React.ReactNode }) {
+  const { isAuthenticated, isLoading } = useAuth();
+  const router = useRouter();
+  const [isChecked, setIsChecked] = useState(false);
```
Is this state necessary? Could you provide more context about it? What does `checked` mean? 🤔

Cema2019 replied:
The ProtectedRoute.tsx component adds the `[isChecked, setIsChecked]` state as an additional `isLoading` check to prevent rendering the component before verifying the token and confirming if the user is already authenticated to proceed with rendering (if authenticated) or redirecting to Home (if not authenticated).

Please confirm if it should be removed or updated.

AlvisonHunterArnuero replied:
The isChecked is been manually controlled by you, so what is the purpose then?

AlvisonHunterArnuero commented on src/utils/api.ts:
```diff
@@ -0,0 +1,30 @@
+
+export type HTTPMethod = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'OPTIONS' | 'HEAD';
```
I suggest we establish a consistent naming convention for custom type aliases to improve code readability. This will make it easier to identify types throughout the codebase. Pluralizing it as well gives a sense of several methods, in my opinion:
```suggestion
export type HttpMethods = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'OPTIONS' | 'HEAD';
```

AlvisonHunterArnuero reviewed (changes requested):
Hi Carlos, Overall, the code looks quite well-structured. However, I do have a couple of suggestions. 
- It might be beneficial to add more comments to the server actions code to explain the logic and error handling process, especially for the token request. 

- Also, consider using a more specific error handling mechanism for the API calls to provide more informative error messages to the users. Additionally, it might be beneficial to add some validation to the input parameters for the server actions to ensure data integrity. 

- Finally, consider adding some unit tests to verify the functionality of the server actions and API calls. Perhaps @Ralexs0096  could give us a hand setting up the test environment on the main branch. 

So, for this PR, all is getting there, good job, @Cema2019

Ralexs0096 commented:
> Hi Carlos, Overall, the code looks quite well-structured. However, I do have a couple of suggestions.
> 
> * It might be beneficial to add more comments to the server actions code to explain the logic and error handling process, especially for the token request.
> * Also, consider using a more specific error handling mechanism for the API calls to provide more informative error messages to the users. Additionally, it might be beneficial to add some validation to the input parameters for the server actions to ensure data integrity.
> * Finally, consider adding some unit tests to verify the functionality of the server actions and API calls. Perhaps @Ralexs0096  could give us a hand setting up the test environment on the main branch.
> 
> So, for this PR, all is getting there, good job, @Cema2019

I will work on the testing config today, @AlvisonHunterArnuero .

Cema2019 committed:
Update src/utils/api.ts

Pluraling method name

Co-authored-by: Alvison Hunter <54810071+AlvisonHunterArnuero@users.noreply.github.com>

Cema2019 committed:
Update src/utils/api.ts

Pluraling method name

Co-authored-by: Alvison Hunter <54810071+AlvisonHunterArnuero@users.noreply.github.com>

Ralexs0096 committed:
WIP

Ralexs0096 committed:
WIP

Ralexs0096 committed:
Refactor authentication check to use fetch instead of fetchAPI

Ralexs0096 committed:
Refactor authentication check to use fetch instead of fetchAPI

Cema2019 committed:
Update logout function without needing to call a new route /auth/logout/ by importing logoutAction from server actions file

Cema2019 committed:
Update logout function without needing to call a new route /auth/logout/ by importing logoutAction from server actions file

Cema2019 committed:
Refactor middleware to enforce redirection for authenticated and non-authenticated users

Cema2019 committed:
Refactor middleware to enforce redirection for authenticated and non-authenticated users

Cema2019 committed:
Fix NEXT_REDIRECT errors in login flow by ensuring intentional redirects aren't caught as errors.

Cema2019 committed:
Fix NEXT_REDIRECT errors in login flow by ensuring intentional redirects aren't caught as errors.

Cema2019 commented:
### Recent Fixes
1. **Fixed token persistence on page refresh thanks to Alex' refactoring**
   - Resolved cookie deletion when refreshing protected routes
   - Implemented proper server-side token verification
   - Ensured client state stays synchronized with authentication status

2. **Complete logout functionality without needing to create a new route /auth/logout**
   - Added server action to delete JWT token
   - Integrated logout with auth context for immediate state update
   - Implemented proper redirect to home page after logout

3. **Route protection enhancements**
   - Middleware now redirects authenticated users away from public routes
   - Protected routes properly validate tokens before rendering

4. **Error handling improvements**
   - Fixed "NEXT_REDIRECT" console errors during login flow
   - Separated actual API errors from intentional redirects

Please review updates and test components.

On page load
![image_2025-06-04_143215942](https://github.com/user-attachments/assets/f2bc9a77-cdbf-492d-9677-7c1280851d19)

After successful login
![image](https://github.com/user-attachments/assets/a6d26180-5625-4b0c-826f-eef070f54b5e)

With Network tab logs
![image](https://github.com/user-attachments/assets/64cbab5d-059b-4a6e-a9e1-a20524c128bc)

After logout
![image](https://github.com/user-attachments/assets/6ce6a685-0ab7-471b-b4cf-3bd680422585)

Ralexs0096 committed:
Adjust Page component after rebase develop changes

Ralexs0096 committed:
Remove unused dependencies from package-lock.json

Cema2019 committed:
Merge branch 'carlos-signin-component' of https://github.com/AlvisonHunterArnuero/dentora-pro into carlos-signin-component

AlvisonHunterArnuero commented:
PWD is being exposed on the network payload, @Cema2019 
![Screenshot 2025-06-06 at 10 36 51 PM](https://github.com/user-attachments/assets/8e4e0b20-134f-4e34-951f-c24f53a63157)

Cema2019 committed:
Remove unnecessary component since route protection is being enforced by middleware

Cema2019 committed:
Remove test Loading component

Cema2019 committed:
Import Elvis component and refactor to match my Express.js server config and cookied-based authentication

Cema2019 committed:
Import and refactor component to display appointments

Cema2019 committed:
Import component to refactor and allow logout functionality

Cema2019 committed:
Refactor component to allow logout and redirection to Home page from dropdown

Cema2019 committed:
Update logout function to redirect to Homepage after deleting token

Cema2019 commented:
## Hello team,

I've implemented recent updates to my component:

- **Import Elvis component and refactor to match my Express.js server config and cookied-based authentication.**
- **Import and refactor component to display appointments.**
- **Import component to refactor and allow logout functionality.**
- **Refactor component to allow logout and redirection to Home page from dropdown.**
- **Update logout function to redirect to Homepage after deleting token.**

Cema2019 committed:
Remove unnecessary creation of formatData object from values

Cema2019 committed:
Destructure email and password variables directly instead of converting to formData

Ralexs0096 commented on src/app/appointments/AppointmentsList.tsx:
```diff
@@ -0,0 +1,156 @@
+        `http://localhost:3001/api/appointments/${appointmentId}`,
+        {
+          method: "DELETE",
+          // body: JSON.stringify({
+          //   appointmentId,
+          // }),
```
Is this code going to be removed?

Cema2019 replied:
Yes, that component will be completely handled and updated by Elvis. 
It was just imported momentarily to replace the original dummy component.

Ralexs0096 commented:
I don't have additional comments. LGTM.

AlvisonHunterArnuero commented:
> I don't have additional comments. LGTM.

Should we proceed with the merge of it, pal? It's your call now...

Cema2019 commented:
Great. 
Thanks for your review. 

Waiting for approval to proceed with PR merge. @Ralexs0096 @AlvisonHunterArnuero

Cema2019 commented:
PR merged upon mentor @Ralexs0096 approval.
